### ISSUE TITLE
[Bug]: AC0032 - false positive

### ISSUE THREAD

**Issue by guidorobben** [ALCops/Analyzers#243]
### Analyzer

ApplicationCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

AC0032

### Description

The rule reports that the table is not used, but as you can see in the code below, this is not true

### AL code to reproduce

```al
 Access = Internal;
    Permissions = 
        tabledata "Purch. Cr. Memo Hdr." = Rm;

    [EventSubscriber(ObjectType::Page, Page::"Pstd. Purch. Cr.Memo - Update", OnAfterRecordChanged, '', false, false)]
    local procedure OnAfterRecordChanged(var PurchCrMemoHdr: Record "Purch. Cr. Memo Hdr."; xPurchCrMemoHdr: Record "Purch. Cr. Memo Hdr."; var IsChanged: Boolean; xPurchCrMemoHdrGlobal: Record "Purch. Cr. Memo Hdr.")
    begin
        IsChanged := IsChanged or
            (PurchCrMemoHdr."Purchaser Code" <> xPurchCrMemoHdrGlobal."Purchaser Code") or
            (PurchCrMemoHdr."Approver WBLC" <> xPurchCrMemoHdrGlobal."Approver WBLC") or
            (PurchCrMemoHdr."On Hold" <> xPurchCrMemoHdrGlobal."On Hold");
    end;

    [EventSubscriber(ObjectType::Codeunit, Codeunit::"Purch. Cr. Memo. Hdr. - Edit", OnBeforePurchCrMemoHdrModify, '', false, false)]
    local procedure OnBeforePurchCrMemoHdrModify(var PurchCrMemoHdr: Record "Purch. Cr. Memo Hdr."; PurchCrMemoHdrRec: Record "Purch. Cr. Memo Hdr.")
    begin
        if PurchCrMemoHdrRec."Status WBLC" = PurchCrMemoHdrRec."Status WBLC"::Released then
            exit;

        if PurchCrMemoHdrRec."Status WBLC" = PurchCrMemoHdrRec."Status WBLC"::"Pending Approval" then begin
            PurchCrMemoHdrRec.CancelApprovalRequestWBLC();
            PurchCrMemoHdr."Status WBLC" := PurchCrMemoHdr."Status WBLC"::Open;
        end;

        PurchCrMemoHdr.UpdateVendorLedgerEntriesWBLC(PurchCrMemoHdr, PurchCrMemoHdrRec);
        PurchCrMemoHdr."Purchaser Code" := PurchCrMemoHdrRec."Purchaser Code";
        PurchCrMemoHdr."Approver WBLC" := PurchCrMemoHdrRec."Approver WBLC";
        PurchCrMemoHdr."On Hold" := PurchCrMemoHdrRec."On Hold";
    end;
```

**Comment by guidorobben** [comment]
v0.8.0 beta3

**Comment by guidorobben** [comment]
<img width="1164" height="70" alt="Image" src="https://github.com/user-attachments/assets/a0dc898e-f347-472b-99de-5441b5cb55c9" />

**Comment by guidorobben** [comment]
I also get a warning for every permission so twice instead of 1

<img width="794" height="48" alt="Image" src="https://github.com/user-attachments/assets/018c71c3-72a0-4ce6-aad9-85b97cbb9640" />

**Comment by Arthurvdv** [comment]
First of thank you for using the beta's releases and providing feedback :-)

The reason the AC0032 raises a diagnostic is that there's no read/insert/modify/delete database operation in the code example you've provided here. I see now the `..no code in this object..` is confusing, so let me update the message that is displayed.

Do you have, by any chance, a small code sample, that shows the message twice? I don't have an repo on this.

**Comment by guidorobben** [comment]
You can use the code provided to see the the double error. I will provide a sample for the false positive

**Comment by guidorobben** [comment]
I also know why this is line is added. If you run the the permissions quickfix from AZ Dev Tools, you get this line. 

<img width="365" height="53" alt="Image" src="https://github.com/user-attachments/assets/db1e6c1b-b8b0-4997-9306-6a2e6dd42c81" />

**Comment by guidorobben** [comment]
This is the false positive

<img width="675" height="293" alt="Image" src="https://github.com/user-attachments/assets/df5a5aaf-4ad6-4e40-a51e-cb2e1b68e574" />

As you can see, there is a Get on the Job table, but also a permission error.

But when I create the sample this is happening

<img width="602" height="204" alt="Image" src="https://github.com/user-attachments/assets/eebc13b7-ba65-47b2-b66c-dd66de30ff51" />

No error anymore. So it looks like a caching issue or something.

**Comment by guidorobben** [comment]
another example of false positive
<img width="687" height="350" alt="Image" src="https://github.com/user-attachments/assets/1834b294-360d-4650-bc8b-9679e9b5c347" />

**Comment by guidorobben** [comment]
Another example of false positive

<img width="933" height="391" alt="Image" src="https://github.com/user-attachments/assets/f192d6d8-860f-4b26-b011-eaafb1e7e541" />

**Comment by Arthurvdv** [comment]
I'm struggling to get a repo on this, while I see in your printscreen more than obviously that these are false positives, I can't repo this on any of my projects. I've also try'ed the BaseApp, but still can't find an repo on this to investigate what is causing this.

**Comment by guidorobben** [comment]
Let me see if I can create a repo I can send you.

**Comment by guidorobben** [comment]
<img width="856" height="582" alt="Image" src="https://github.com/user-attachments/assets/6f69659a-b64d-4202-8901-6a45eceeafcb" />

You can download it from this location
https://github.com/guidorobben/BluaceWorkflowEditor

**Comment by Arthurvdv** [comment]
Thanks, I've downloaded the example and let's see if I can find the culprit...

**Comment by Arthurvdv** [comment]
I've try'ed with AL Language version `17.0.2273547` and `18.0.2293710` on a Windows VM (Standard_D4s_v5) with only the AL Language extension and ALCops installed, where I can't get to reproduce your case.

Do you maybe any idea what I can try?

<img width="1398" height="840" alt="Image" src="https://github.com/user-attachments/assets/30a880f9-b407-4b5e-80e6-ede54ad932ce" />

**Comment by guidorobben** [comment]
That is funny, In my screenshot you can see that "Purch. Inv. Header " has a warning too. I have 17.0.2273547 running.

hmm. You have the same settings file so all the same code analyzers are enabled. Same ruleset. 

What version of ALCops are you using?

**Comment by guidorobben** [comment]
When disabling ApplicationCop, the warnings are gone, so it is at least related to ALCops. I tried turning off different extensions but so far no luck.

**Comment by Arthurvdv** [comment]
Yeah, is absolutely the ApplicationCop raising these AC0032 diagnostics, but still now idea on why this behaviour is different then I would expect.

Can you try with the attached [ALCops.ApplicationCop.dll.zip](https://github.com/user-attachments/files/27369916/ALCops.ApplicationCop.dll.zip) to unzip and manually place this into the `%userprofile%\.vscode\extensions\ms-dynamics-smb.al-17.0.2273547\bin\Analyzers\` and see if this raises any AC9999 diagnostics?

**Comment by guidorobben** [comment]
Nope, AC9999 is not raised, only AC0032

**Comment by guidorobben** [comment]
<img width="3238" height="785" alt="Image" src="https://github.com/user-attachments/assets/d2e4e22f-ba4a-477b-bc20-9cd7a881a972" />

**Comment by Arthurvdv** [comment]
That MissingMethodException is on me, I should have also send you the updated ALCops.Common.dll.

I did some rework on the rule in `v0.8.0-beta.4`, where maybe the bug is gone but I'm unsure as I'm still unable to find a repo so I'm still a bit in the dark. If this doesn't work I'll maybe disable the rule by default for now and release v0.8.0 with this rule disabled and look into this for the next release.

**Comment by guidorobben** [comment]
Still the same with 0.8.0 beta 4. am I the only one with this symptom?

**Comment by guidorobben** [comment]
something else is going on.

this is the initial, see the last 2 dont have a warning
<img width="673" height="214" alt="Image" src="https://github.com/user-attachments/assets/666b5438-436a-4c94-89ad-045f10e13dd3" />

when I delete the first  2,  it now marks the last2.

<img width="643" height="171" alt="Image" src="https://github.com/user-attachments/assets/19fa93b2-f965-478b-a4f3-0c80c297c86f" />

Looks like some caching is going on.

**Comment by Arthurvdv** [comment]
I wish I could replicate this somehow. I'm thinking out loud here, is this maybe a multi-root workspace instead of a single project opened?

**Comment by guidorobben** [comment]
Could be. 90% of the time I have a workspace open. 

But I had the Workflow Editor open as a single project.

**Comment by Arthurvdv** [comment]
> am I the only one with this symptom?

You're one the brave ones with me on the beta channel, I unfortunately didn't get any other reports in of this. I did some more rework and I've changed some inner workings for `v0.8.0-beta.5`, which hopefully improves the rule.

**Comment by Arthurvdv** [comment]
I've released this new rule into the wild with the release of version `v0.8.0` of ALCops.

I will keep this issue open for now to see if the problem of the false positives is resolved (fingers crossed)

**Comment by tvwijk** [comment]
I get a false positive when using a report:

<img width="1760" height="34" alt="Image" src="https://github.com/user-attachments/assets/c58b4643-928f-4b5f-a9b7-7281acef507f" />

<img width="512" height="42" alt="Image" src="https://github.com/user-attachments/assets/fca0df8e-95d4-41d4-b77d-ef5d90c3dfeb" />

<img width="410" height="305" alt="Image" src="https://github.com/user-attachments/assets/a0bb2d6e-6243-483b-8d31-b146a96605bc" />

It does not take a variable used in a dataitem into consideration.

**Comment by AnsBarMIT** [comment]
I also get this, see example below (in a workspace with multiple extensions if relevant):

<img width="918" height="396" alt="Image" src="https://github.com/user-attachments/assets/7343ff71-68b2-4bfc-9118-b54bd37125e4" />

```al
report 53003 Test
{
    UsageCategory = ReportsAndAnalysis;
    ApplicationArea = All;
    ProcessingOnly = true;
    Permissions = tabledata Customer = rm; // Permission 'm' on 'tabledata MITCCommissionWorksheetLine' is not needed. Only 'r' is required.

    dataset
    {
        dataitem(Customer; Customer)
        {
            trigger OnAfterGetRecord()
            begin
                Customer.Name := 'test';
                Customer.Modify(true);
            end;
        }
    }
}
```

**Comment by Arthurvdv** [comment]
Thanks @tvwijk and @AnsBarMIT for reporting the false positives on reports.

This should be fixed with the upcoming release of v0.8.1 of ALCops.

**Comment by Arthurvdv** [comment]
With the release of version `0.8.1` of ALCops, the false positives should now be resolved.

Since this issue has grown to include a variety of different scenarios and edge cases, I’m going to close it to help keep things manageable. If you still run into a false positive after updating, please feel free to open a new issue with the details. That would help me better track remaining cases.
